File: glyph/src/main/java/com/hsj86715/ingress/glyph/BaseDecoration.java
package com.hsj86715.ingress.glyph;

import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;

/**
 * Created by hushujun on 2017/5/16.
 */

public class BaseDecoration extends RecyclerView.ItemDecoration {
    private static final String TAG = "BaseDecoration";

    protected boolean isFirstColumn(@NonNull RecyclerView.LayoutManager layoutManager, int position) {
        final int spanCount;
        final int orientation;
        final int itemCount = layoutManager.getItemCount();
        if (layoutManager instanceof GridLayoutManager) {
            spanCount = ((GridLayoutManager) layoutManager).getSpanCount();
            orientation = ((GridLayoutManager) layoutManager).getOrientation();
            if (orientation == GridLayoutManager.VERTICAL) {
                return position % spanCount == 0;
            } else {
                final int more = itemCount % spanCount;
                if (more == 0) {
                    return position % spanCount == 0;
                } else {
                    final int row = position / spanCount;
                    if (row < more) {
                        return position % spanCount == 0;
                    } else {
                        return (position + row - more) % spanCount == 0;
                    }
                }
            }
        } else if (layoutManager instanceof LinearLayoutManager) {
            orientation = ((LinearLayoutManager) layoutManager).getOrientation();
            if (orientation == LinearLayoutManager.VERTICAL) {
                return true;
            } else {
                return position == 0;
            }
        } else {
            Log.i(TAG, "isFirstColumn==>the LayoutManager is not support," + layoutManager);
        }
        return false;
    }

    protected boolean isLastColumn(@NonNull RecyclerView.LayoutManager layoutManager, int position) {
        final int spanCount;
        final int orientation;
        final int itemCount = layoutManager.getItemCount();
        if (layoutManager instanceof GridLayoutManager) {
            spanCount = ((GridLayoutManager) layoutManager).getSpanCount();
            orientation = ((GridLayoutManager) layoutManager).getOrientation();
            if (orientation == GridLayoutManager.VERTICAL) {
                return position % spanCount == (spanCount - 1) || position == itemCount - 1;
            } else {
                final int more = itemCount % spanCount;
                if (more == 0) {
                    return position % spanCount == spanCount - 1;
                } else {
                    final int row = position / spanCount;
                    if (row < more) {
                        return position % spanCount == spanCount - 1;
                    } else {
                        return (position + row - more) % spanCount == spanCount - 1;
                    }
                }
            }
        } else if (layoutManager instanceof LinearLayoutManager) {
            orientation = ((LinearLayoutManager) layoutManager).getOrientation();
            if (orientation == LinearLayoutManager.VERTICAL) {
                return true;
            } else {
                return position == itemCount - 1;
            }
        } else {
            Log.i(TAG, "isLastColumn==>the LayoutManager is not support," + layoutManager);
        }
        return false;
    }

    protected boolean isFirstRow(@NonNull RecyclerView.LayoutManager layoutManager, int position) {
        final int spanCount;
        final int orientation;
        final int itemCount = layoutManager.getItemCount();
        if (layoutManager instanceof GridLayoutManager) {
            spanCount = ((GridLayoutManager) layoutManager).getSpanCount();
            orientation = ((GridLayoutManager) layoutManager).getOrientation();
            if (orientation == GridLayoutManager.VERTICAL) {
                return position < spanCount;
            } else {
                final int more = itemCount % spanCount;
                if (more == 0) {
                    return position < itemCount / spanCount;
                } else {
                    return position < itemCount / spanCount + 1;
                }
            }
        } else if (layoutManager instanceof LinearLayoutManager) {
            orientation = ((LinearLayoutManager) layoutManager).getOrientation();
            if (orientation == LinearLayoutManager.VERTICAL) {
                return position == 0;
            } else {
                return true;
            }
        } else {
            Log.i(TAG, "isFirstRow==>the LayoutManager is not support," + layoutManager);
        }
        return false;
    }

    protected boolean isLastRow(@NonNull RecyclerView.LayoutManager layoutManager, int position) {
        final int spanCount;
        final int orientation;
        final int itemCount = layoutManager.getItemCount();
        if (layoutManager instanceof GridLayoutManager) {
            spanCount = ((GridLayoutManager) layoutManager).getSpanCount();
            orientation = ((GridLayoutManager) layoutManager).getOrientation();
            if (orientation == GridLayoutManager.VERTICAL) {
                return position >= itemCount - spanCount;
            } else {
                int more = itemCount % spanCount;
                int row = position / spanCount + 1;
                return row == spanCount || spanCount * more < position;
            }
        } else if (layoutManager instanceof LinearLayoutManager) {
            orientation = ((LinearLayoutManager) layoutManager).getOrientation();
            if (orientation == LinearLayoutManager.VERTICAL) {
                return position == itemCount - 1;
            } else {
                return true;
            }
        } else {
            Log.i(TAG, "isLastRow==>the LayoutManager is not support," + layoutManager);
        }
        return false;
    }
}

File: app/src/main/java/com/hsj86715/ingress/glyph/BaseDecoration.kt
package com.hsj86715.ingress.glyph

import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.util.Log

/**
 * Created by hushujun on 2017/4/10.
 */
open class BaseDecoration : RecyclerView.ItemDecoration() {
    val TAG: String = "BaseDecoration"

    fun isFirstColumn(layoutManager: RecyclerView.LayoutManager, position: Int): Boolean {
        val spanCount: Int
        val orientation: Int
        val itemCount: Int = layoutManager.itemCount
        if (layoutManager is GridLayoutManager) {
            spanCount = layoutManager.spanCount
            orientation = layoutManager.orientation
            if (orientation == GridLayoutManager.VERTICAL) {
                return position % spanCount == 0
            } else {
                val more = itemCount % spanCount
                if (more == 0) {
                    return position % spanCount == 0
                } else {
                    val row = position / spanCount
                    if (row < more) {
                        return position % spanCount == 0
                    } else {
                        return (position + row - more) % spanCount == 0
                    }
                }
            }
        } else if (layoutManager is LinearLayoutManager) {
            orientation = layoutManager.orientation
            if (orientation == LinearLayoutManager.VERTICAL) {
                return true
            } else {
                return position == 0
            }
        } else if (layoutManager is StaggeredGridLayoutManager) {
            Log.i(TAG, "isFirstColumn==>mLayoutManager is StaggeredGridLayoutManager")
            // TODO: 2017/4/17
        }
        return false
    }

    fun isLastColumn(layoutManager: RecyclerView.LayoutManager, position: Int): Boolean {
        val spanCount: Int
        val orientation: Int
        val itemCount: Int = layoutManager.itemCount
        if (layoutManager is GridLayoutManager) {
            spanCount = layoutManager.spanCount
            orientation = layoutManager.orientation
            if (orientation == GridLayoutManager.VERTICAL) {
                return position % spanCount == (spanCount - 1) || position == itemCount - 1
            } else {
                val more = itemCount % spanCount
                if (more == 0) {
                    return position % spanCount == spanCount - 1
                } else {
                    val row = position / spanCount
                    if (row < more) {
                        return position % spanCount == spanCount - 1
                    } else {
                        return (position + row - more) % spanCount == spanCount - 1
                    }
                }
            }
        } else if (layoutManager is LinearLayoutManager) {
            orientation = layoutManager.orientation
            if (orientation == LinearLayoutManager.VERTICAL) {
                return true
            } else {
                return position == itemCount - 1
            }
        } else if (layoutManager is StaggeredGridLayoutManager) {
            Log.i(TAG, "isLastColumn==>mLayoutManager is StaggeredGridLayoutManager")
            // TODO: 2017/4/17
//            spanCount = mLayoutManager.spanCount
//            orientation = mLayoutManager.orientation
//            if (orientation == StaggeredGridLayoutManager.VERTICAL) {
//
//            } else {
//
//            }
        }
        return false
    }

    fun isFirstRow(layoutManager: RecyclerView.LayoutManager, position: Int): Boolean {
        val spanCount: Int
        val orientation: Int
        val itemCount: Int = layoutManager.itemCount
        if (layoutManager is GridLayoutManager) {
            spanCount = layoutManager.spanCount
            orientation = layoutManager.orientation
            if (orientation == GridLayoutManager.VERTICAL) {
                return position < spanCount
            } else {
                val more = itemCount % spanCount
                if (more == 0) {
                    return position < itemCount / spanCount
                } else {
                    return position < itemCount / spanCount + 1
                }
            }
        } else if (layoutManager is LinearLayoutManager) {
            orientation = layoutManager.orientation
            if (orientation == LinearLayoutManager.VERTICAL) {
                return position == 0
            } else {
                return true
            }
        } else if (layoutManager is StaggeredGridLayoutManager) {
            Log.i(TAG, "isFirstRow==>mLayoutManager is StaggeredGridLayoutManager")
            // TODO: 2017/4/17
        }
        return false
    }

    fun isLastRow(layoutManager: RecyclerView.LayoutManager, position: Int): Boolean {
        val spanCount: Int
        val orientation: Int
        val itemCount: Int = layoutManager.itemCount
        if (layoutManager is GridLayoutManager) {
            spanCount = layoutManager.spanCount
            orientation = layoutManager.orientation
            if (orientation == GridLayoutManager.VERTICAL) {
                return position >= itemCount - spanCount
            } else {
                val more = itemCount % spanCount
                val row = position / spanCount + 1
                return row == spanCount || spanCount * more < position
            }
        } else if (layoutManager is LinearLayoutManager) {
            orientation = layoutManager.orientation
            if (orientation == LinearLayoutManager.VERTICAL) {
                return position == itemCount - 1
            } else {
                return true
            }
        } else if (layoutManager is StaggeredGridLayoutManager) {
            Log.i(TAG, "isLastRow==>mLayoutManager is StaggeredGridLayoutManager")
            // TODO: 2017/4/17
        }
        return false
    }
}